src/modules/sdexec/recover.c

```diff
@@ -0,0 +1,117 @@
+ * Copyright 2023 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* recover.c - reset/stop all units with our unit name prefix
+ *
+ * Later we'll actually try to recover running jobs, hence the name.
+ * For now just make sure any lingering processes are cleaned up.
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <stdio.h>
+#include <jansson.h>
+#include <flux/core.h>
+
+#include "ccan/str/str.h"
+#include "src/common/libutil/errno_safe.h"
+
+#include "stop.h"
+#include "recover.h"
+
+struct unit {
+    const char *name;
+    const char *desc;
+    const char *load;
+    const char *active;
+    const char *sub;
+};
+
+static int sdexec_parse_unit (json_t *o, struct unit *unit)
+{
+    if (json_unpack (o,
+                     "[sssss]",
+                     &unit->name,
+                     &unit->desc,
+                     &unit->load,
+                     &unit->active,
+                     &unit->sub) < 0) {
+        errno = EPROTO;
+        return -1;
+    }
+    return 0;
+}
+
+static int sdexec_list_units (flux_t *h, const char *pattern, json_t **op)
+{
+    flux_future_t *f;
+    json_t *units;
+
+    if (!(f = flux_rpc_pack (h,
+                             "sdbus.call",
+                             FLUX_NODEID_ANY,
+                             0,
+                             "{s:s s:[[] [s]]}",
+                             "member", "ListUnitsByPatterns",
+                             "params", pattern))
+        || flux_rpc_get_unpack (f,
+                                "{s:[o]}",
+                                "params", &units) < 0) {
+        flux_log (h,
+                  LOG_ERR,
+                  "ListUnitsByPattern: %s",
+                  future_strerror (f, errno));
+        flux_future_destroy (f);
+        return -1;
+    }
+    *op = json_incref (units);
+    flux_future_destroy (f);
+    return 0;
+}
+
+int sdexec_recover (flux_t *h, const char *unit_prefix)
+{
+    char *pattern;
+    json_t *units = NULL;
+    size_t index;
+    json_t *entry;
+    int rc = -1;
+
+    if (asprintf (&pattern, "%s*.service", unit_prefix) < 0) {
+        flux_log (h, LOG_ERR, "recover: out of memory");
+        return -1;
+    }
+    if (sdexec_list_units (h, pattern, &units) < 0)
+        goto out;
+    json_array_foreach (units, index, entry) {
+        struct unit unit;
+        flux_future_t *f;
+
+        if (sdexec_parse_unit (entry, &unit) < 0) {
+            flux_log (h, LOG_ERR, "recover: error parsing unit object");
+            goto out;
+        }
+        if (streq (unit.active, "failed"))
+            f = sdexec_reset_failed_unit (h, unit.name);
+        else
+            f = sdexec_stop_unit (h, unit.name, "fail");
```

Comment:
in `libsdprocess` i basically have logic of

```
        if (streq (unit.active, "active")
            f = sdexec_stop_unit (h, unit.name, "fail");
        else if (streq (unit.active, "failed"))
            f = sdexec_reset_failed_unit (h, unit.name);
        else
            /* error unexpected state */
```

I recall on rare occasions getting other states, which I presumed was due to something racy, like perhaps catching a unit at that exact moment it's between two states we care about.  And IIRC reset/fail and stop unit may not work.  Most notably is the use of RemainAfterExit, b/c I think a state can go from inactive back to active.

But your logic might be better, b/c you are doing 

```
        if (streq (unit.active, "failed"))
            f = sdexec_reset_failed_unit (h, unit.name);
        else
            f = sdexec_stop_unit (h, unit.name, "fail");
```

i.e. do stop if not failed, and that might work for all states != "failed".

Not sure if you've seen issues here, but perhaps worthwhile to atleast log if the active state is something other than what we expect??

Reply:
I'll ponder that one.  I think your idea of documenting/logging all the state changes might help this become clearer.